checker_c#/N0t3b00kChecker/NotebookClient.cs

```diff
@@ -0,0 +1,313 @@
+        private TcpClient? tcpClient;
+        private StreamReader? reader;
+        private StreamWriter? writer;
+
+        public NotebookClient(ILogger<NotebookClient> logger)
+        {
+            this.logger = logger;
+        }
+
+        public async Task Connect(CheckerTaskMessage task, CancellationToken token)
+        {
+            this.logger.LogInformation("Connecting to service");
+            this.tcpClient = new();
+            try
+            {
+                await this.tcpClient.ConnectAsync(task.Address, Checker.SERVICE_PORT, token);
+                this.reader = new StreamReader(this.tcpClient.GetStream());
+                this.writer = new StreamWriter(this.tcpClient.GetStream());
+                token.Register(() =>
+                {
+                    tcpClient.Close();
+                    reader.Close();
+                    writer.Close();
+                });
+            }
+            catch (Exception e)
+            {
+                this.logger.LogWarning($"Connect failed: {e.Message}\n{e.StackTrace}");
+                throw new OfflineException("Failed to establish TCP connection");
+            }
+
+            string? response;
+            try
+            {
+                this.logger.LogDebug("Wait for welcome message");
+                response = await this.reader!.ReadLineAsync();
+                await reader!.ReadAsync(new char[2].AsMemory(), token);
+            }
+            catch (Exception e)
+            {
+                this.logger.LogWarning($"Received no welcome message: {e.Message}\n{e.StackTrace}");
+                throw new OfflineException("No welcome message received");
+            }
+
+            if (response != "Welcome to the 1337 n0t3b00k!")
+            {
+                this.logger.LogWarning($"Received no welcome message: {response}");
+                throw new MumbleException("No welcome message received");
+            }
+        }
+
+        public async Task Register(NotebookUser user, CancellationToken token)
+        {
+            this.logger.LogInformation($"Registering user {user.Username}");
+            this.user = user;
+            try
+            {
+                this.logger.LogDebug($"reg {this.user.Username} {this.user.Password}");
+                await this.writer!.WriteAsync($"reg {this.user.Username} {this.user.Password}\n".AsMemory(), token);
+                await this.writer.FlushAsync();
+            }
+            catch (Exception e)
+            {
+                this.logger.LogWarning($"Register failed: {e.Message}\n{e.StackTrace}");
+                throw new OfflineException("Connection error during registration");
+            }
+
+            string? response;
+            try
+            {
+                this.logger.LogDebug("Wait for register success message");
+                response = await this.reader!.ReadLineAsync();
+                await reader!.ReadAsync(new char[2].AsMemory(), token);

```

Comment:
Result of `ReadAsync` needs to be checked (or better yet, find, or add a library function that reads the expected amount of bytes)

Reply:
The example python checker doesn't do that correctly too /shrug

Reply:
The python checker reads YOLO amount of bytes (which may be wrong for the service? cc @gehaxelt ).
This line looks like it tries to read exactly two and does not check the return. If that's what you intended, fine, but students will copy and reuse this code without knowing it can fail.

Reply:
And it does not read the whitespace after the `>`, so I assumed it is fair to ignore a "broken" prompt of the correct length

Reply:
My point is that `ReadAsync` should _always_ be checked. If you are fine with a smaller return value, do so explicitly, and add a comment, so students can understand and learn.

Reply:
Ideally, add a `ReadAsyncExact` method to the checkerlib that will loop until all bytes have been read, or throw Mumble- or Offline-exceptions accordingly